File: src/api/registry.ts
/**
 * Agent Pack Registry API client — Story 12.5 AC7, AC3, AC6
 */

import { getApiBase } from './_base';
const API_BASE_URL = getApiBase();

export class RegistryApiError extends Error {
  constructor(
    public status: number,
    public detail: unknown,
    public code: string = '',
  ) {
    super(`Registry API error ${status}`);
  }
}

function _extractCode(detail: unknown): string {
  if (detail && typeof detail === 'object') {
    const err = (detail as Record<string, unknown>).error;
    if (err && typeof err === 'object') {
      const code = (err as Record<string, unknown>).code;
      if (typeof code === 'string') return code;
    }
  }
  return '';
}

async function _handle<T>(res: Response): Promise<T> {
  if (res.ok) return res.json() as Promise<T>;
  const body = await res.text();
  let detail: unknown = body;
  try { detail = JSON.parse(body); } catch { /* raw */ }
  throw new RegistryApiError(res.status, detail, _extractCode(detail));
}

interface Envelope<T> {
  data: T;
  meta: Record<string, unknown>;
}

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface PackRecord {
  id: string;
  version: string;
  name: string;
  description: string;
  author: string;
  tags: string[];
  capabilities_summary: string[];
  install_status: 'not_installed' | 'installed' | 'has_update';
  verified: boolean;
}

export interface InstalledPackRecord {
  pack_id: string;
  pack_version: string;
  agent_id: string;
  name: string;
  installed_at: string;
  update_available: boolean;
  verified: boolean;
}

export interface InstallResult {
  agent_id: string;
  pack_id: string;
  pack_version: string;
  blueprint: Record<string, unknown>;
  installed_at: string;
  verified: boolean;
  already_installed: boolean;
}

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

export async function listPacks(params?: {
  workspace_id?: string;
  tags?: string;
  q?: string;
}): Promise<PackRecord[]> {
  // 2026-05-11 fix — `new URL('/api/...')` (no host) throws TypeError. Use
  // URLSearchParams for the query, concat with template string.
  const sp = new URLSearchParams();
  if (params?.workspace_id) sp.set('workspace_id', params.workspace_id);
  if (params?.tags) sp.set('tags', params.tags);
  if (params?.q) sp.set('q', params.q);
  const qs = sp.toString();
  const env = await _handle<Envelope<PackRecord[]>>(
    await fetch(`${API_BASE_URL}/api/agents/registry/packs${qs ? '?' + qs : ''}`),
  );
  return env.data;
}

export async function installPack(packId: string, workspaceId = 'default'): Promise<InstallResult> {
  const res = await fetch(`${API_BASE_URL}/api/agents/registry/packs/install`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ pack_id: packId, workspace_id: workspaceId }),
  });
  const env = await _handle<Envelope<InstallResult>>(res);
  return env.data;
}

export async function listInstalledPacks(workspaceId = 'default'): Promise<InstalledPackRecord[]> {
  const env = await _handle<Envelope<InstalledPackRecord[]>>(
    await fetch(`${API_BASE_URL}/api/agents/registry/packs/installed?workspace_id=${encodeURIComponent(workspaceId)}`),
  );
  return env.data;
}

File: shadowflow/api/registry.py
"""Agent Pack Registry API — Story 12.5 AC3, AC4, AC6, AC7.

Prefix: /api/agents/registry/packs  (list + install + installed)

Note: /api/agents/registry (ACP runtime registry) is already handled in
`shadowflow/api/agents.py` — this module adds Pack-level endpoints under
a longer prefix to avoid collision.
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Dict, List, Optional
from uuid import uuid4

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel, Field

from shadowflow.services.registry_service import get_registry_service

router = APIRouter(prefix="/api/agents/registry/packs", tags=["agent-pack-registry"])


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _envelope(data: Any, **meta: Any) -> Dict[str, Any]:
    return {"data": data, "meta": meta}


def _http_error(status: int, code: str, message: str, **extra: Any) -> HTTPException:
    return HTTPException(
        status_code=status,
        detail={"error": {"code": code, "message": message, **extra}},
    )


import re as _re
_WORKSPACE_ID_RE = _re.compile(r"^[a-zA-Z0-9_-]+$")


def _validate_workspace_id(workspace_id: str) -> None:
    """Raise 400 if workspace_id contains characters outside [a-zA-Z0-9_-]."""
    if not _WORKSPACE_ID_RE.match(workspace_id):
        raise _http_error(
            400,
            "INVALID_WORKSPACE_ID",
            "workspace_id must match [a-zA-Z0-9_-]+",
        )


# ---------------------------------------------------------------------------
# Request bodies
# ---------------------------------------------------------------------------


class InstallRequest(BaseModel):
    pack_id: str = Field(..., min_length=1, max_length=120)
    workspace_id: str = Field(
        "default",
        min_length=1,
        max_length=80,
        pattern=r"^[a-zA-Z0-9_-]+$",
    )


# ---------------------------------------------------------------------------
# Routes
# ---------------------------------------------------------------------------


@router.get("")
async def list_packs(
    workspace_id: str = Query("default"),
    tags: Optional[str] = Query(None, description="Comma-separated tags to filter"),
    q: Optional[str] = Query(None, description="Keyword search in name/description"),
) -> Dict[str, Any]:
    """List all available Agent Packs with install status."""
    _validate_workspace_id(workspace_id)
    svc = get_registry_service()
    entries = svc.list_manifests(tags=tags, q=q)
    result: List[Dict[str, Any]] = []
    for entry in entries:
        manifest = svc.get_manifest(entry.id)
        if manifest is None:
            continue
        rec = svc.get_installed_record(entry.id, workspace_id)
        verified = rec.get("verified", False) if rec else manifest.signature is not None
        result.append({
            "id": manifest.id,
            "version": manifest.version,
            "name": manifest.name,
            "description": manifest.description,
            "author": manifest.author,
            "tags": entry.tags,
            "capabilities_summary": manifest.capabilities.tools,
            "install_status": svc.install_status(manifest.id, workspace_id),
            "verified": verified,
        })
    return _envelope(result, total=len(result))


@router.post("/install", status_code=201)
async def install_pack(body: InstallRequest) -> Dict[str, Any]:
    """Install an Agent Pack — creates an AgentBlueprint record."""
    svc = get_registry_service()

    # Step 1 — find pack
    manifest = svc.get_manifest(body.pack_id)
    if manifest is None:
        raise _http_error(404, "PACK_NOT_FOUND", f"Pack '{body.pack_id}' not found in registry")

    # Step 2 — verify signature (AC5)
    # `verified` = True only when a signature is PRESENT *and* passes HMAC check.
    verified = False
    if manifest.signature is not None:
        if not svc.verify_signature(manifest):
            raise _http_error(
                400,
                "MANIFEST_SIGNATURE_INVALID",
                "Manifest signature verification failed",
                pack_id=body.pack_id,
            )
        verified = True  # signature present and verified successfully

    # Step 3 — already installed same version?
    installed_ver = svc.get_installed_version(body.pack_id, body.workspace_id)
    if installed_ver == manifest.version:
        rec = svc.get_installed_record(body.pack_id, body.workspace_id)
        return _envelope(
            {**(rec or {}), "already_installed": True},
            warnings=[],
            deprecations=[],
        )

    # Step 4 — install_cmd
    # Subprocess execution is intentionally deferred to a post-MVP story (no
    # sandboxing / privilege model exists yet).  For now we surface a warning
    # in the response so callers know the hook was NOT run.
    install_cmd_warning: Optional[str] = None
    if manifest.install_cmd:
        install_cmd_warning = (
            f"install_cmd '{manifest.install_cmd}' was NOT executed — "
            "automatic hook execution is not yet supported. Run it manually."
        )

    # Step 5 — create AgentBlueprint-compatible record
    agent_id = f"agent-{uuid4().hex[:12]}"
    now = datetime.now(timezone.utc).isoformat()
    blueprint = {
        "name": manifest.name,
        "goal": manifest.soul,
        "role_profiles": [
            {
                "name": manifest.name,
                "description": manifest.description,
                "persona": manifest.soul,
                "tools": manifest.capabilities.tools,
                "executor_kind": manifest.kind,
                "executor_provider": manifest.capabilities.llm_provider,
            }
        ],
        "tool_policies": [
            {"tool_id": t, "default_permission": "allow"}
            for t in manifest.capabilities.tools
        ],
        "metadata": {
            "source": "catalog",
            "pack_id": manifest.id,
            "pack_version": manifest.version,
            "verified": verified,
        },
    }

    # Step 6 — persist installed-packs record (AC4 storage schema)
    install_rec: Dict[str, Any] = {
        "pack_id": manifest.id,
        "pack_version": manifest.version,
        "agent_id": agent_id,
        "installed_at": now,
        "verified": verified,
        "blueprint": blueprint,
    }
    try:
        svc.upsert_installed(body.workspace_id, install_rec)
    except Exception as exc:
        raise _http_error(500, "PERSISTENCE_FAILED", f"Persistence failed: {exc}") from exc

    response_warnings: List[str] = []
    if not verified:
        response_warnings.append("Pack has no signature — treat as community/unverified")
    if install_cmd_warning:
        response_warnings.append(install_cmd_warning)

    return _envelope(
        {
            "agent_id": agent_id,
            "pack_id": manifest.id,
            "pack_version": manifest.version,
            "blueprint": blueprint,
            "installed_at": now,
            "verified": verified,
            "already_installed": False,
        },
        warnings=response_warnings,
        deprecations=[],
    )


@router.get("/installed")
async def list_installed(
    workspace_id: str = Query("default"),
) -> Dict[str, Any]:
    """List installed Agent Packs for a workspace."""
    _validate_workspace_id(workspace_id)
    svc = get_registry_service()
    items = svc.list_installed(workspace_id)
    return _envelope(items, total=len(items))
